File: ainpc-core-plugin/src/test/java/ro/ainpc/debug/WorldMappingSemanticIndexTest.java
package ro.ainpc.debug;

import org.junit.jupiter.api.Test;
import ro.ainpc.world.PlaceType;
import ro.ainpc.world.StoryMode;
import ro.ainpc.world.WorldNodeInfo;
import ro.ainpc.world.WorldPlaceInfo;
import ro.ainpc.world.WorldRegionInfo;

import java.util.List;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

class WorldMappingSemanticIndexTest {

    @Test
    void indexesMarketPlaceAndQuestBoardNodeForMappedContracts() {
        WorldRegionInfo region = new WorldRegionInfo(
            "demo_sat",
            "Demo Sat",
            "world",
            "settlement",
            0,
            50,
            0,
            100,
            90,
            100,
            List.of("village", "public"),
            StoryMode.EVOLUTIVE,
            "default",
            List.of()
        );
        WorldPlaceInfo market = new WorldPlaceInfo(
            "demo_sat:piata",
            "demo_sat",
            "Piata satului",
            "world",
            PlaceType.MARKET,
            20,
            60,
            20,
            40,
            75,
            40,
            List.of("market", "public", "social"),
            "",
            true,
            Map.of("role", "social")
        );
        WorldNodeInfo board = new WorldNodeInfo(
            "demo_sat:piata:quest_board",
            "demo_sat",
            "demo_sat:piata",
            "quest_trigger",
            "world",
            30.0,
            65.0,
            30.0,
            2.0,
            Map.of("semantic", "quest_board")
        );

        WorldMappingSemanticIndex index = WorldMappingSemanticIndex.from(
            List.of(region),
            List.of(market),
            List.of(board)
        );

        assertEquals(List.of("demo_sat:piata"), index.placeTags().get("market"));
        assertEquals(List.of("demo_sat:piata"), index.placeTypes().get("market"));
        assertEquals(List.of("demo_sat:piata"), index.placeCandidates().get("market"));
        assertEquals(List.of("demo_sat:piata:quest_board"), index.nodeTypes().get("quest_trigger"));
        assertEquals(List.of("demo_sat:piata:quest_board"), index.nodeMetadataValues().get("semantic_quest_board"));
        assertEquals(List.of("demo_sat:piata:quest_board"), index.nodeCandidates().get("quest_board"));
        assertTrue(index.regionCandidates().get("settlement").contains("demo_sat"));
        assertTrue(index.hasReference("place", "tag:market"));
        assertTrue(index.hasReference("node", "quest_board"));
        assertEquals(List.of("demo_sat:piata"), index.matchingIds("place", "type:market"));
        assertEquals(List.of("demo_sat:piata:quest_board"), index.matchingIds("node", "type:quest_trigger"));
        assertTrue(index.matchingIds("place", "tag:castle").isEmpty());
    }
}

File: ainpc-core-plugin/bin/test/ro/ainpc/debug/WorldMappingSemanticIndexTest.kt
package ro.ainpc.debug

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import ro.ainpc.world.PlaceType
import ro.ainpc.world.StoryMode
import ro.ainpc.world.WorldNodeInfo
import ro.ainpc.world.WorldPlaceInfo
import ro.ainpc.world.WorldRegionInfo

class WorldMappingSemanticIndexTest {
    @Test
    fun indexesMarketPlaceAndQuestBoardNodeForMappedContracts() {
        val region = WorldRegionInfo(
            "demo_sat", "Demo Sat", "world", "settlement",
            0, 50, 0, 100, 90, 100,
            listOf("village", "public"), StoryMode.EVOLUTIVE, "default", listOf()
        )
        val market = WorldPlaceInfo(
            "demo_sat:piata", "demo_sat", "Piata satului", "world", PlaceType.MARKET,
            20, 60, 20, 40, 75, 40,
            listOf("market", "public", "social"), "", true, mapOf("role" to "social")
        )
        val board = WorldNodeInfo(
            "demo_sat:piata:quest_board", "demo_sat", "demo_sat:piata", "quest_trigger", "world",
            30.0, 65.0, 30.0, 2.0, mapOf("semantic" to "quest_board")
        )

        val index = WorldMappingSemanticIndex.from(listOf(region), listOf(market), listOf(board))

        assertEquals(listOf("demo_sat:piata"), index.placeTags()["market"])
        assertEquals(listOf("demo_sat:piata"), index.placeTypes()["market"])
        assertEquals(listOf("demo_sat:piata"), index.placeCandidates()["market"])
        assertEquals(listOf("demo_sat:piata:quest_board"), index.nodeTypes()["quest_trigger"])
        assertEquals(listOf("demo_sat:piata:quest_board"), index.nodeMetadataValues()["semantic_quest_board"])
        assertEquals(listOf("demo_sat:piata:quest_board"), index.nodeCandidates()["quest_board"])
        assertTrue(index.regionCandidates()["settlement"]!!.contains("demo_sat"))
        assertTrue(index.hasReference("place", "tag:market"))
        assertTrue(index.hasReference("node", "quest_board"))
        assertEquals(listOf("demo_sat:piata"), index.matchingIds("place", "type:market"))
        assertEquals(listOf("demo_sat:piata:quest_board"), index.matchingIds("node", "type:quest_trigger"))
        assertTrue(index.matchingIds("place", "tag:castle").isEmpty())
    }
}
